File: web/src/api/cloud/instance.js
import service from '@/utils/request'
// @Tags Instance
// @Summary 创建实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body model.Instance true "创建实例管理"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"创建成功"}"
// @Router /inst/createInstance [post]
export const createInstance = (data) => {
  return service({
    url: '/inst/createInstance',
    method: 'post',
    data
  })
}

// @Tags Instance
// @Summary 删除实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body model.Instance true "删除实例管理"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"删除成功"}"
// @Router /inst/deleteInstance [delete]
export const deleteInstance = (params) => {
  return service({
    url: '/inst/deleteInstance',
    method: 'delete',
    params
  })
}

// @Tags Instance
// @Summary 批量删除实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body request.IdsReq true "批量删除实例管理"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"删除成功"}"
// @Router /inst/deleteInstance [delete]
export const deleteInstanceByIds = (params) => {
  return service({
    url: '/inst/deleteInstanceByIds',
    method: 'delete',
    params
  })
}

// @Tags Instance
// @Summary 更新实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body model.Instance true "更新实例管理"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"更新成功"}"
// @Router /inst/updateInstance [put]
export const updateInstance = (data) => {
  return service({
    url: '/inst/updateInstance',
    method: 'put',
    data
  })
}

// @Tags Instance
// @Summary 用id查询实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data query model.Instance true "用id查询实例管理"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"查询成功"}"
// @Router /inst/findInstance [get]
export const findInstance = (params) => {
  return service({
    url: '/inst/findInstance',
    method: 'get',
    params
  })
}

// @Tags Instance
// @Summary 分页获取实例管理列表
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data query request.PageInfo true "分页获取实例管理列表"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /inst/getInstanceList [get]
export const getInstanceList = (params) => {
  return service({
    url: '/inst/getInstanceList',
    method: 'get',
    params
  })
}
// @Tags Instance
// @Summary 获取数据源
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Success 200 {string} string "{"success":true,"data":{},"msg":"查询成功"}"
// @Router /inst/findInstanceDataSource [get]
export const getInstanceDataSource = () => {
  return service({
    url: '/inst/getInstanceDataSource',
    method: 'get',
  })
}

// @Tags Instance
// @Summary 不需要鉴权的实例管理接口
// @Accept application/json
// @Produce application/json
// @Param data query cloudReq.InstanceSearch true "分页获取实例管理列表"
// @Success 200 {object} response.Response{data=object,msg=string} "获取成功"
// @Router /inst/getInstancePublic [get]
export const getInstancePublic = () => {
  return service({
    url: '/inst/getInstancePublic',
    method: 'get',
  })
}

export const closeInstance = (data) => {
  return service({
    url: '/inst/closeInstance',
    method: 'post',
    data
  })
}

export const restartInstance = (data) => {
  return service({
    url: '/inst/restartInstance',
    method: 'post',
    data
  })
}

File: server/api/v1/cloud/instance.go
package cloud

import (
	"github.com/flipped-aurora/gin-vue-admin/server/global"
	"github.com/flipped-aurora/gin-vue-admin/server/model/cloud"
	cloudReq "github.com/flipped-aurora/gin-vue-admin/server/model/cloud/request"
	"github.com/flipped-aurora/gin-vue-admin/server/model/common/response"
	"github.com/flipped-aurora/gin-vue-admin/server/utils"
	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

type InstanceApi struct{}

// CreateInstance 创建实例管理
// @Tags Instance
// @Summary 创建实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body cloud.Instance true "创建实例管理"
// @Success 200 {object} response.Response{msg=string} "创建成功"
// @Router /inst/createInstance [post]
func (instApi *InstanceApi) CreateInstance(c *gin.Context) {
	// 创建业务用Context
	ctx := c.Request.Context()

	var inst cloud.Instance
	err := c.ShouldBindJSON(&inst)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	userID := utils.GetUserID(c)
	inst.UserID = new(int64)
	*inst.UserID = int64(userID)

	err = instService.CreateInstance(ctx, &inst)
	if err != nil {
		global.GVA_LOG.Error("创建失败!", zap.Error(err))
		response.FailWithMessage("创建失败:"+err.Error(), c)
		return
	}
	response.OkWithMessage("创建成功", c)
}

// CloseInstance 关闭实例
// @Tags Instance
// @Summary 关闭实例
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body cloud.Instance true "关闭实例"
// @Success 200 {object} response.Response{msg=string} "关闭成功"
// @Router /inst/closeInstance [post]
func (instApi *InstanceApi) CloseInstance(c *gin.Context) {
	var inst cloud.Instance
	err := c.ShouldBindJSON(&inst)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = instService.CloseInstance(c.Request.Context(), &inst)
	if err != nil {
		global.GVA_LOG.Error("关闭失败!", zap.Error(err))
		response.FailWithMessage("关闭失败:"+err.Error(), c)
		return
	}
	response.OkWithMessage("关闭成功", c)
}

// RestartInstance 重启实例
// @Tags Instance
// @Summary 重启实例
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body cloud.Instance true "重启实例"
// @Success 200 {object} response.Response{msg=string} "重启成功"
// @Router /inst/restartInstance [post]
func (instApi *InstanceApi) RestartInstance(c *gin.Context) {
	var inst cloud.Instance
	err := c.ShouldBindJSON(&inst)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = instService.RestartInstance(c.Request.Context(), &inst)
	if err != nil {
		global.GVA_LOG.Error("重启失败!", zap.Error(err))
		response.FailWithMessage("重启失败:"+err.Error(), c)
		return
	}
	response.OkWithMessage("重启成功", c)
}

// DeleteInstance 删除实例管理
// @Tags Instance
// @Summary 删除实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body cloud.Instance true "删除实例管理"
// @Success 200 {object} response.Response{msg=string} "删除成功"
// @Router /inst/deleteInstance [delete]
func (instApi *InstanceApi) DeleteInstance(c *gin.Context) {
	// 创建业务用Context
	ctx := c.Request.Context()

	ID := c.Query("ID")
	err := instService.DeleteInstance(ctx, ID)
	if err != nil {
		global.GVA_LOG.Error("删除失败!", zap.Error(err))
		response.FailWithMessage("删除失败:"+err.Error(), c)
		return
	}
	response.OkWithMessage("删除成功", c)
}

// DeleteInstanceByIds 批量删除实例管理
// @Tags Instance
// @Summary 批量删除实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Success 200 {object} response.Response{msg=string} "批量删除成功"
// @Router /inst/deleteInstanceByIds [delete]
func (instApi *InstanceApi) DeleteInstanceByIds(c *gin.Context) {
	// 创建业务用Context
	ctx := c.Request.Context()

	IDs := c.QueryArray("IDs[]")
	err := instService.DeleteInstanceByIds(ctx, IDs)
	if err != nil {
		global.GVA_LOG.Error("批量删除失败!", zap.Error(err))
		response.FailWithMessage("批量删除失败:"+err.Error(), c)
		return
	}
	response.OkWithMessage("批量删除成功", c)
}

// UpdateInstance 更新实例管理
// @Tags Instance
// @Summary 更新实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data body cloud.Instance true "更新实例管理"
// @Success 200 {object} response.Response{msg=string} "更新成功"
// @Router /inst/updateInstance [put]
func (instApi *InstanceApi) UpdateInstance(c *gin.Context) {
	// 从ctx获取标准context进行业务行为
	ctx := c.Request.Context()

	var inst cloud.Instance
	err := c.ShouldBindJSON(&inst)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	err = instService.UpdateInstance(ctx, inst)
	if err != nil {
		global.GVA_LOG.Error("更新失败!", zap.Error(err))
		response.FailWithMessage("更新失败:"+err.Error(), c)
		return
	}
	response.OkWithMessage("更新成功", c)
}

// FindInstance 用id查询实例管理
// @Tags Instance
// @Summary 用id查询实例管理
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param ID query uint true "用id查询实例管理"
// @Success 200 {object} response.Response{data=cloud.Instance,msg=string} "查询成功"
// @Router /inst/findInstance [get]
func (instApi *InstanceApi) FindInstance(c *gin.Context) {
	// 创建业务用Context
	ctx := c.Request.Context()

	ID := c.Query("ID")
	reinst, err := instService.GetInstance(ctx, ID)
	if err != nil {
		global.GVA_LOG.Error("查询失败!", zap.Error(err))
		response.FailWithMessage("查询失败:"+err.Error(), c)
		return
	}
	response.OkWithData(reinst, c)
}

// GetInstanceList 分页获取实例管理列表
// @Tags Instance
// @Summary 分页获取实例管理列表
// @Security ApiKeyAuth
// @Accept application/json
// @Produce application/json
// @Param data query cloudReq.InstanceSearch true "分页获取实例管理列表"
// @Success 200 {object} response.Response{data=response.PageResult,msg=string} "获取成功"
// @Router /inst/getInstanceList [get]
func (instApi *InstanceApi) GetInstanceList(c *gin.Context) {
	// 创建业务用Context
	ctx := c.Request.Context()

	var pageInfo cloudReq.InstanceSearch
	err := c.ShouldBindQuery(&pageInfo)
	if err != nil {
		response.FailWithMessage(err.Error(), c)
		return
	}
	list, total, err := instService.GetInstanceInfoList(ctx, pageInfo)
	if err != nil {
		global.GVA_LOG.Error("获取失败!", zap.Error(err))
		response.FailWithMessage("获取失败:"+err.Error(), c)
		return
	}
	response.OkWithDetailed(response.PageResult{
		List:     list,
		Total:    total,
		Page:     pageInfo.Page,
		PageSize: pageInfo.PageSize,
	}, "获取成功", c)
}

// GetInstanceDataSource 获取Instance的数据源
// @Tags Instance
// @Summary 获取Instance的数据源
// @Accept application/json
// @Produce application/json
// @Success 200 {object} response.Response{data=object,msg=string} "查询成功"
// @Router /inst/getInstanceDataSource [get]
func (instApi *InstanceApi) GetInstanceDataSource(c *gin.Context) {
	// 创建业务用Context
	ctx := c.Request.Context()

	// 此接口为获取数据源定义的数据
	dataSource, err := instService.GetInstanceDataSource(ctx)
	if err != nil {
		global.GVA_LOG.Error("查询失败!", zap.Error(err))
		response.FailWithMessage("查询失败:"+err.Error(), c)
		return
	}
	response.OkWithData(dataSource, c)
}

// GetInstancePublic 不需要鉴权的实例管理接口
// @Tags Instance
// @Summary 不需要鉴权的实例管理接口
// @Accept application/json
// @Produce application/json
// @Success 200 {object} response.Response{data=object,msg=string} "获取成功"
// @Router /inst/getInstancePublic [get]
func (instApi *InstanceApi) GetInstancePublic(c *gin.Context) {
	// 创建业务用Context
	ctx := c.Request.Context()

	// 此接口不需要鉴权
	// 示例为返回了一个固定的消息接口，一般本接口用于C端服务，需要自己实现业务逻辑
	instService.GetInstancePublic(ctx)
	response.OkWithDetailed(gin.H{
		"info": "不需要鉴权的实例管理接口信息",
	}, "获取成功", c)
}
